testing-guidelines/ci.md

```diff
@@ -0,0 +1,431 @@
+
+Workflows are a discreet set of jobs with a discrete set of steps. It might be reasonable to assume that a workflow is a workflow. Sadly this isn't true, the event that triggers the workflow dictates not only when that workflow will run (which makes sense) but also a bunch of other information about both its environment and even which version of that workflow file will run (this is a git repo after all). This latter feature, _also_ makes sense, but it isn't immediately apparent.
+
+- `pull_request` - This event runs correctly on contributor PRs and check out the correct branch by default (more on this later) but it doesn't not have access to secrets.
+- `pull_request_target` - Same as `pull_request` but it _does_ have access to secrets. However because this event runs in to context of the branch (and repo) the PR is made from, that PR has also has direct access to secrets making it insecure.
+- `push` - When triggered from a fork, this will essentially be a `push` to the fork not the target repo. In many cases the workflow won't run ata ll (depends on the settings of the fork) and it won't have access to secrets from the base repo. Even if it did, it would be inscure.
+
+There are ways to run workflows indirectly:
+
+- `workflow_dispatch` - This event always runs in the context of main  You can programatically trigger this workflow event, allowing more control over where that workflow runs but you need to use the GitHub API to do this. Ttherefore the triggering workflow needs access to secrets, rendering it insecure for our purposes.
+- `workflow_run` - This is essentially `workflow_dispatch` inverted. Instead of triggering it from elsewhere explicitly, the workflow _itself_ determines which workflow will trigger _it_. This means that you do not need access to secrets in order to start a `workflow_run` and since this event type runs in the context of main, it is secure.
+
+<details>
+  <summary>What does the "context" of an event mean?</summary>
+
+In GitHub Actions 'context' is a somewhat overloaded term, but it typically describes the permissions, available data, and the source code state that a given workflow has access to, without any additional code. 
+
+For example, you can check out any branch of any public repo in any workflow but the context is important before any configured steps are run, in fact, the context is important before the workflow even starts. For practical purposes, there are two elements to the 'context' that people care about.
+
+- Which workflow file on which branch actually runs.
+- What information about the repo or triggering event does that workflow have access to.
+
+If a workflow "runs in the context of the default branch" then it will use the workflow that exists on trhe default branch, regardless of whether or not the event that originall triggered it was on another branch. If the workflow "runs in the context of the pull request branch" then it will pull the workflow file from the pull request branch.
```

Comment:
```suggestion
If a workflow "runs in the context of the default branch" then it will use the workflow that exists on the default branch, regardless of whether or not the event that originally triggered it was on another branch. If the workflow "runs in the context of the pull request branch" then it will pull the workflow file from the pull request branch.
```